binary_search_tree/BinarySearchTree.java

```diff
@@ -0,0 +1,175 @@
+import java.util.NoSuchElementException;
+
+class Node {	// Node for Binary Search Tree
+	public int data;
+	public Node left;
+	public Node right;
+
+	public Node(int data) {	// Initializes node with given data and no chlid
+		this.data = data;
+		this.left = null;
+		this.right = null;
+	}
+}
+
+class BST {
+	private Node root;	// Root of Binary Search Tree
+
+	public BST() {	// Initializes empty BST
+		root = null;
+	}
+
+	public BST(int data) {	// Initializes root of BST
+		root = new Node(data);
+	}
+
+	public void insert(int data) {
+		if (root == null) {
+			root = new Node(data);	// Initialize root with the given data
+			return;
+		}
+		Node newNode = new Node(data);	// Create Node for new entry
+		Node iterator = root;	// Temp Node for iterating from root to leaf
+		Node parent = null;	// To store the future parent of new node
+		while (iterator != null) {
+			parent = iterator;
+			if (data <= iterator.data) {
+				iterator = iterator.left;
+			}
+			else {
+				iterator = iterator.right;
+			}
+		}
+		if (data <= parent.data) {
+			parent.left = newNode;
+		}
+		else {
+			parent.right = newNode;
+		}
+	}
+
+	public Node search(int data) {	// Returns the Node if element is found else will throw an Exception
+		Node iterator = root;
+		while (iterator != null) {
+			if (iterator.data == data)
+				return iterator;
+			else if (data <= iterator.data)
+				iterator = iterator.left;
+			else
+				iterator = iterator.right;
+		}
+		throw new NoSuchElementException("Element is not found in BST");
+	}
+
+	public boolean delete(int data) {	// Finds the parent of the node to be deleted.
+		if (root == null) {
+			throw new NoSuchElementException("Cannot perform delete operation, BST is empty");
+		}
+		Node iterator = root;
+		Node parent = null;
+		while (iterator != null) {
+			parent = iterator;
+			if (data == iterator.data)
+				return deleteNode(data, parent);
+			else if (data <= iterator.data)
+				iterator = iterator.left;
+			else
+				iterator = iterator.right;
+		}
+		throw new NoSuchElementException("Delete Unsuccessful! Element was not found in BST");
```

Comment:
@mohitkyadav I've catched the exception inside `main()` and also mentioned there in comment that it can be tested for the other one too. However there seem to be some other error with my code in delete operation. I will test try-catch block with delete operation too as you have requested.